relay-server/src/services/buffer/mod.rs

```diff
@@ -81,6 +84,44 @@ impl FromMessage<Self> for EnvelopeBuffer {
     }
 }
 
+/// Abstraction that wraps a list of [`ObservableEnvelopeBuffer`]s to which [`Envelope`] are routed
+/// based on their [`ProjectKeyPair`].
+#[derive(Debug, Clone)]
+pub struct PartitionedEnvelopeBuffer {
+    buffers: Arc<Vec<ObservableEnvelopeBuffer>>,
+    partitions: u32,
+}
+
+impl PartitionedEnvelopeBuffer {
+    /// Creates a new [`PartitionedEnvelopeBuffer`] with already instantiated buffers and the number
+    /// of partitions.
+    pub fn new(buffers: Vec<ObservableEnvelopeBuffer>, partitions: u32) -> Self {
+        debug_assert!(buffers.len() == partitions as usize);
+        Self {
+            buffers: Arc::new(buffers),
+            partitions,
+        }
+    }
+
+    /// Returns the [`ObservableEnvelopeBuffer`] to which [`Envelope`]s having the supplied
+    /// [`ProjectKeyPair`] will be sent.
+    ///
+    /// The rationale of using this partitioning strategy is to reduce memory usage across buffers
+    /// since each individual buffer will only take care of a subset of projects.
+    pub fn buffer(&self, project_key_pair: ProjectKeyPair) -> Option<&ObservableEnvelopeBuffer> {
+        let mut hasher = DefaultHasher::new();
```

Comment:
I'm not sure whether `DefaultHasher` returns consistent results across restarts. At least it doesn't guarantee that across releases: https://doc.rust-lang.org/std/collections/hash_map/struct.DefaultHasher.html

Can we use `FnvHash`? It's simple and already in our dependencies.

Reply:
This is a very good point, it would not cause any functional issues but it would definitely make the system less memory efficient given the lack of locality.